rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,430 @@
+jarring to users, not to mention that potentially having a copy of each atomic
+memory operation with a `_volatile` prefix would get annoying quickly when
+reading through the API docs.
+
+Also, it is extremely common to want either all operations on a memory
+location to be volatile, or none of them. Providing separate wrapper types
+helps enforce this very common usage pattern at the API level.
+
+## Self-type or not self-type
+[self-type]: #self-type
+
+Instead of using `arbitrary_self_types` to get `a.load(Ordering::Relaxed)`
+method syntax on pointer-like objects, one could instead provide the volatile
+operations as inherent methods, i.e. `VolatileU8::load(a, Ordering::Relaxed)`.
+
+This has the advantage of avoiding coupling this feature to another unstable
+feature. But it has the drawback of being incredibly verbose. Typing the same
+volatile type name over and over again in a complex shared memory transaction
+would certainly get old and turn annoying quickly, and we don't want anger to
+distract low-level developers from the delicate task of implementing the kind
+of subtle algorithm that requires volatile operations.
+
+## `NonNull<T>` vs `*mut T` vs `*const T` vs other
+[pointers]: #pointers
+
+Honestly, I don't have a very strong opinion there. I have a positive _a priori_
+towards `NonNull<T>` because it encodes an important invariant in the API, and
+I think that arbitrary self types make its usually poor ergonomics bearable.
+
+But what I would really want here is something like a non-`dereferenceable`
+Rust reference. I don't like the fact that I have to give up on the borrow
+checker and make everything unsafe just to do volatile loads and stores, which
+are not unsafe _per se_ as long as they occur via a `Volatile` wrapper. It just
+feels like we should be able to propose something better than full references
+vs full raw pointers here...
+
+## Full atomics vocabulary vs hardware semantics
+[atomics-vs-hardware]: #atomics-vs-hardware
+
+Currently, this RFC basically proposes exposing a volatile version of every
+atomic operation supported by Rust for maximal expressive power, and I could
+definitely find uses for the new possibilities that this opens in IPC scenarios.
+
+But it could also be argued that this distracts us from volatile's main purpose
+of generating a stream of simple hardware instructions without using inline
+assembly:
+
+- Non-`Relaxed` atomics will entail memory barriers
+- Compare-and-swap may be implemented as a load-linked/store-conditional loop
+- Some types like `VolatileBool` are dangerous when interacting with untrusted
+  code because they come with data validity invariants.
+
+From this perspective, there would be an argument in favor of only supporting
+`Relaxed` load/stores and machine data types, at least initially. And I could
+get behind that. But since the rest can be useful in IPC with trusted Rust code,
+I thought it might be worth at least considering.
```

Comment:
It would be interesting to know to what do e.g. volatile SeqCst loads / store / cas expand to on x86, arm, power, etc.

Reply:
You may be interested in [this web page](https://www.cl.cam.ac.uk/~pes20/cpp/cpp0xmappings.html). Not sure if the RFC should link to it though, as ultimately this particular implementation choice in in the hands of LLVM, not rustc...

Reply:
That link is very useful. I was interested in knowing which volatile atomic operations would expand to more than one instruction, and that answered it. Thanks!